MaxText/layers/deepseek.py

```diff
@@ -0,0 +1,213 @@
+  )
+
+  attention_lnx = attention_layer(
+      lnx,
+      lnx,
+      decoder_positions,
+      decoder_segment_ids=decoder_segment_ids,
+      deterministic=deterministic,
+      model_mode=model_mode,
+  )
+
+  attention_lnx = nn.with_logical_constraint(
+      attention_lnx, ("activation_batch", "activation_norm_length", "activation_embed")
+  )
+  intermediate_inputs = inputs + attention_lnx
+
+  # Normalization
+  hidden_states = models.RMSNorm(
+      dtype=cfg.dtype,
+      weight_dtype=cfg.weight_dtype,
+      name="post_self_attention_layer_norm",
+      kernel_axes=("norm",),
+      epsilon=cfg.normalization_layer_epsilon,
+  )(intermediate_inputs)
+  hidden_states = nn.with_logical_constraint(
+      hidden_states, ("activation_batch", "activation_norm_length", "activation_embed")
+  )
+  return hidden_states, intermediate_inputs
+
+
+def post_process(cfg, layer_output):
+  if cfg.record_internal_nn_metrics:
+    self.sow("intermediates", "activation_mean", jnp.mean(layer_output))
+    self.sow("intermediates", "activation_stdev", jnp.std(layer_output))
+    self.sow(
+        "intermediates",
+        "activation_fraction_zero",
+        jnp.sum(layer_output == 0) / jnp.size(layer_output),
+    )
+
+  if cfg.scan_layers:
+    return layer_output, None
+  else:
+    return layer_output
+
+
+class DeepSeekDenseLayer(nn.Module):
+  """DeepSeek-style dense layer."""
+
+  config: models.Config
+  mesh: Mesh
+  quant: Optional[Quant] = None
+
+  @nn.compact
+  def __call__(
+      self,
+      inputs,
+      decoder_segment_ids,
+      decoder_positions,
+      deterministic,
+      model_mode,
+  ):
+    cfg = self.config
+    inputs = nn.with_logical_constraint(inputs, ("activation_batch", "activation_norm_length", "activation_embed"))
+    inputs = checkpoint_name(inputs, "decoder_layer_input")
+
+    hidden_states, intermediate_inputs = self_attention_with_norm(
+        inputs, cfg, self.mesh, self.quant, decoder_segment_ids, decoder_positions, deterministic, model_mode
+    )
+    mlp_lnx = linears.MlpBlock(
+        intermediate_dim=cfg.mlp_dim,
+        activations=cfg.mlp_activations,
+        intermediate_dropout_rate=cfg.dropout_rate,
+        dtype=cfg.dtype,
+        weight_dtype=cfg.weight_dtype,
+        name="mlp",
+        config=cfg,
+        quant=self.quant,
+    )(hidden_states, deterministic=deterministic)
+    mlp_lnx = nn.with_logical_constraint(mlp_lnx, ("activation_batch", "activation_norm_length", "activation_embed"))
+
+    layer_output = mlp_lnx + intermediate_inputs
+    layer_output = nn.Dropout(rate=cfg.dropout_rate, broadcast_dims=(-2,))(layer_output, deterministic=deterministic)
+    layer_output = nn.with_logical_constraint(
+        layer_output,
+        ("activation_batch", "activation_norm_length", "activation_embed"),
+    )
+    return post_process(self.config, layer_output)
+
+
+class DeepSeekMoELayer(nn.Module):
+  """DeepSeek-style MoE layer."""
```

Comment:
Can you comment what are main differences of DeepSeekMoELayer from regular MoELayer?

Reply:
Sure, will add!